crypto/threads_common.c

```diff
@@ -0,0 +1,451 @@
+ *        Pointer to the thread-local `SPARSE_ARRAY_OF(MASTER_KEY_ENTRY)`
+ *        structure to be cleaned up.
+ */
+static void clean_master_key(void *data)
+{
+    SPARSE_ARRAY_OF(MASTER_KEY_ENTRY) *mkey = data;
+
+    ossl_sa_MASTER_KEY_ENTRY_doall_arg(mkey, clean_master_key_id, NULL);
+    ossl_sa_MASTER_KEY_ENTRY_free(mkey);
+}
+
+/**
+ * @brief Initializes the thread-local storage for master key data.
+ *
+ * This function sets up the thread-local key used to store per-thread
+ * master key tables. It also registers the `clean_master_key` function
+ * as the destructor to be called when the thread exits.
+ *
+ * This function is intended to be called once using `CRYPTO_THREAD_run_once`
+ * to ensure thread-safe initialization.
+ */
+static void init_master_key(void)
+{
+    /*
+     * Note: We assign a cleanup function here, which is atypical for
+     * uses of CRYPTO_THREAD_init_local.  This is because, nominally
+     * we expect that the use of ossl_init_thread_start will be used
+     * to notify openssl of exiting threads.  However, in this case
+     * we want the metadata for this interface (the sparse arrays) to
+     * stay valid until the thread actually exits, which is what the
+     * clean_master_key function does.  Data held in the sparse arrays
+     * (that is assigned via CRYPTO_THREAD_set_local_ex), are still expected
+     * to be cleaned via the ossl_init_thread_start/stop api.
+     */
+    CRYPTO_THREAD_init_local(&master_key, clean_master_key);
+
+    /*
+     * Indicate that the key has been set up.
+     */
+    master_key_init = 1;
+}
+
+/**
+ * @brief Retrieves context-specific data from thread-local storage.
+ *
+ * This function looks up and returns the data associated with a given
+ * thread-local key ID and `OSSL_LIB_CTX` context. The data must have
+ * previously been stored using `CRYPTO_THREAD_set_local_ex()`.
+ *
+ * If the master key table is not yet initialized, it will be lazily
+ * initialized via `init_master_key()`. If the requested key or context
+ * entry does not exist, `NULL` is returned.
+ *
+ * @param id
+ *        The thread-local key ID used to identify the master key entry.
+ *
+ * @param ctx
+ *        Pointer to the `OSSL_LIB_CTX` used to index into the context
+ *        table for the specified key.
+ *
+ * @return A pointer to the stored context-specific data, or NULL if no
+ *         entry is found or initialization fails.
+ */
+void *CRYPTO_THREAD_get_local_ex(CRYPTO_THREAD_LOCAL_KEY_ID id, OSSL_LIB_CTX *ctx)
+{
+    SPARSE_ARRAY_OF(MASTER_KEY_ENTRY) *mkey;
+    MASTER_KEY_ENTRY *entry;
+    CTX_TABLE_ENTRY *ctxd;
+
+    /*
+     * Make sure the master key has been initalized
+     * NOTE: We use CRYPTO_THREAD_run_once here, rather than the
+     * RUN_ONCE macros.  We do this because this code is included both in
+     * libcrypto, and in fips.[dll|dylib|so].  FIPS attempts to avoid doing
+     * one time initalization of global data, and so supresses the definition
```

Comment:
As an aside: The reason we had this restriction was because we had no easy way of cleaning up global data at the time (it can't be cleaned up on provider tear down, because the provider might be loaded into multiple libctxs). So we made it a rule that everything had to go into the libctx (no global data that needs to be cleaned up). However we actually do have a way of doing this now in the self_test `cleanup` function